### PR TITLE
#3

### PR DESCRIPTION
**Pull request by koleCar**

### CODE REVIEW

File: client/shared/components/confirmation/confirmation.component.html [hunk 1]
```diff
@@ -12,7 +12,7 @@ <h1 mat-dialog-title>{{options.header}}</h1>
   <button
     mat-raised-button
     tabindex="-1"
-    color="warn"
+    color="primary"
```
Comment (flauc) [inline client/shared/components/confirmation/confirmation.component.html]: I think we need to make this dynamic. The thing is, the confirmation is used a lot for confirming a removal. In which case `color="warn"` is appropriate. On the guard, though we would indeed need it as `color="primary"`

File: client/shared/guards/can-deactivate.guard.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            description:
+              'You made changes on this page. What would you like to do with them before you leave?',
+            confirm: 'Continue editing',
+            negate: 'Discard changes'
+          }
+        })
+        .afterClosed()
+        .pipe(
+          map(res => {
+            console.log('res', res);
```
Comment (flauc) [inline client/shared/guards/can-deactivate.guard.ts]: This is what is breaking the build `console.log` isn't allowed in the source code.

File: client/apps/dashboard/src/app/pages/products/products.module.ts [hunk 3]
```diff
@@ -6,14 +6,19 @@ import {ProductsListComponent} from './pages/list/products-list.component';
...
   {
     path: '',
     component: ProductsComponent,
     children: [
       {path: '', component: ProductsListComponent},
-      {path: ':id', component: ProductsSinglePageComponent},
+      {
+        path: ':id',
+        component: ProductsSinglePageComponent,
+        canDeactivate: [CanDeactivateGuard]
```
Comment (flauc) [inline client/apps/dashboard/src/app/pages/products/products.module.ts]: We need to incorporate this on all the other single pages as well.

File: client/apps/dashboard/src/app/pages/products/pages/single-page/products-single-page.component.ts [hunk 4]
```diff
@@ -28,6 +28,8 @@ export class ProductsSinglePageComponent extends LangSinglePageComponent
 
   categories$: Observable<Category[]>;
   collection = FirestoreCollections.Products;
+  initialValue: any;
```
Comment (flauc) [inline client/apps/dashboard/src/app/pages/products/pages/single-page/products-single-page.component.ts]: This implementation should be done on the `SinglePageComponent`.

File: client/apps/dashboard/src/app/pages/categories/single-page/categories-single-page.component.ts [hunk 5]
```diff
@@ -21,5 +21,7 @@ export class CategoriesSinglePageComponent extends LangSinglePageComponent {
       name: [data.name || '', Validators.required],
       description: [data.description || '']
     });
+
+    this.connectGuard();
```
Comment (flauc) [inline client/apps/dashboard/src/app/pages/categories/single-page/categories-single-page.component.ts]: I don't think we need to call this on every component. Couldn't we call it in the `SinglePageComponent`?

File: client/shared/components/confirmation/confirmation.component.html [hunk 6]
```diff
@@ -12,7 +12,7 @@ <h1 mat-dialog-title>{{options.header}}</h1>
   <button
     mat-raised-button
     tabindex="-1"
-    color="warn"
+    color="{{options.color}}"
```
Comment (flauc) [inline client/shared/components/confirmation/confirmation.component.html]: Let's use the `[]` whenever possible

File: client/shared/components/confirmation/confirmation-options.interface.ts [hunk 7]
```diff
@@ -3,4 +3,5 @@ export interface ConfirmationOptions {
   negate: string;
   header: string;
   description?: string;
+  color?: string;
```
Comment (flauc) [inline client/shared/components/confirmation/confirmation-options.interface.ts]: This isn't a string. There is only a limited number of valid values `warn`, `primary` and `default`. I'm pretty sure angular material exports an enum with those values. If they don't though. Let's make our own.

File: client/apps/dashboard/src/app/shared/components/lang-single-page/lang-single-page.component.ts [hunk 8]
```diff
@@ -11,6 +11,9 @@ import {SinglePageComponent} from '../single-page/single-page.component';
 })
 export class LangSinglePageComponent extends SinglePageComponent
   implements OnInit {
+  initialValue: any;
+  currentValue: any;
```
Comment (flauc) [inline client/apps/dashboard/src/app/shared/components/lang-single-page/lang-single-page.component.ts]: I think we should define this on the `SinglePageComponent` that way we can use it on the customer and order for example.

File: client/shared/guards/can-deactivate.guard.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+import {ConfirmationComponent} from '@jf/components/confirmation/confirmation.component';
+import {map} from 'rxjs/operators';
+import {of} from 'rxjs';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class CanDeactivateGuard implements CanDeactivate<any> {
+  constructor(private _dialog: MatDialog) {}
+  initialKey = 'initialValue';
```
Comment (flauc) [inline client/shared/guards/can-deactivate.guard.ts]: I don't think this is necessary. Can't we just do `component.initialValue` and `component.currentValue` below?

File: client/apps/dashboard/src/app/pages/categories/single-page/categories-single-page.component.ts (hunk 5 above)
Comment (koleCar) [reply client/apps/dashboard/src/app/pages/categories/single-page/categories-single-page.component.ts]: Every build form function is defined in it's own component, so after we populate data in form we need to call this function.
I cant find a way to remove this from component

**Pull request merged**